Blish HUD/GameServices/ArcDpsService.cs

```diff
@@ -1,96 +1,122 @@
 ﻿using System;
 using System.Diagnostics;
-using System.IO.Pipes;
+using System.Net;
 using System.Threading;
-using System.Threading.Tasks;
+using Blish_HUD.ArcDps;
+using Blish_HUD.ArcDps.Models;
 using Microsoft.Xna.Framework;
-using ProtoBuf;
 
 namespace Blish_HUD
 {
     public class ArcDpsService : GameService
     {
-        public bool ArcPresent;
-        private CancellationToken Cancel;
-        public bool HudIsActive;
-        private Mutex Mut;
-        private NamedPipeServerStream Server;
-        private Stopwatch Watch;
+        public delegate void RawCombatEvent(CombatEvent data);
+
+        private static readonly object WatchLock = new object();
+        private readonly TimeSpan _leeway = TimeSpan.FromMilliseconds(1000);
+        private bool _hudIsActive;
+#if DEBUG
+        public static long Counter = 0;
+#endif
+
+        private SocketListener _server;
+
+        private Stopwatch _stopwatch;
+
+        /// <summary>
+        ///     Indicates if arcdps updated <see cref="HudIsActive" /> in the last second (it should every in-game frame)
+        /// </summary>
+        public bool RenderPresent { get; private set; }
+
+        /// <summary>
+        ///     Indicates if arcdps currently draws its HUD (not in character select, cut scenes or loading screens)
+        /// </summary>
+        public bool HudIsActive
+        {
+            get
+            {
+                lock (WatchLock)
+                {
+                    return _hudIsActive;
+                }
+            }
+            private set
+            {
+                lock (WatchLock)
+                {
+                    _stopwatch.Restart();
+                    _hudIsActive = value;
+                }
+            }
+        }
+
+        /// <summary>
+        ///     Holds unprocessed combat data
+        /// </summary>
+        public event RawCombatEvent OnRawCombatEvent;
 
         protected override void Initialize()
         {
-            Server = new NamedPipeServerStream("BHUDrender");
-            Mut = new Mutex(false);
+            _stopwatch = new Stopwatch();
+            _server = new SocketListener(10, 200_000);
+            _server.ReceivedMessage += MessageHandler;
+#if DEBUG
+            OnRawCombatEvent += data =>
+            {
+                Interlocked.Increment(ref Counter);
+            };
+#endif
         }
 
         protected override void Load()
         {
-            Cancel = new CancellationToken(false);
-            Watch = Stopwatch.StartNew();
-            Task.Run(() =>
-            {
-                while (true)
-                    try
-                    {
-                        Server.WaitForConnection();
-                        var obj = Serializer.Deserialize<Arc>(Server);
-                        Server.Disconnect();
+            _server.Start(new IPEndPoint(IPAddress.Loopback, 8214));
```

Comment:
Was port 8214 selected randomly?  This is pedantic and while I doubt it would be an issue, I think best practice would be to select an ephemeral port from the range 49152-65535 that's open from the ArcDps plugin and then provide someway for the service to determine that port (either via registry, saved file, etc.).

Seems like it'd be a lot of trouble for the chance to save maybe one edge case, so probably not worth doing.  I think at the very least, though, a port from a higher range (say 5 digits long) would make me feel better.  :)

Reply:
8 21 4
h u d
i just checked with the wikipedia article of commonly used ports and selected something thats free
i dont like port determination, because it adds points of failure. if you have a really good idea, then convince me though. the whole point of this application is ipc anyway ^^

Reply:
Okay, well for the clever choice, I'll have to give you this one!